File: backend/routers/calidad.py
from fastapi import APIRouter, Query
import sqlite3
import pandas as pd
from typing import List, Optional

router = APIRouter(prefix="/api/calidad")
DB_PATH = "/Users/danielramirezquintana/Desktop/TOK3M_1/DB/TOKEM.db"

COLS = {
    "SALUDO": "EvCal_Saludo", "TITULAR": "EvCal_Titular", "FAMILIAR": "EvCal_Familiar",
    "PRESENTACION": "EvCal_Presentacion", "CORDIALIDAD": "EvCal_Cordialidad",
    "RECADO": "EvCal_Recado", "EMPEX": "EvCal_Empex", "ENCARGO": "EvCal_Encargo",
    "GRABADO": "EvCal_Grabado", "INFORMACION": "EvCal_Informacion",
    "MOTIVO": "EvCal_Motivo", "OFERTA": "EvCal_Oferta", "CANALES": "EvCal_Canales",
    "COPA": "EvCal_Copa", "DUDAS": "EvCal_Dudas", "CIERRE": "EvCal_Cierre"
}

def get_where(inicio, fin, codigos, empresas, ejecutivos):
    f = []
    if inicio and fin:
        f.append(f"YMD BETWEEN '{inicio.replace('-', '')}' AND '{fin.replace('-', '')}'")
    if codigos:
        vals = ", ".join(["'" + str(c) + "'" for c in codigos])
        f.append(f"CODIGO_CONTACTO IN ({vals})")
    if empresas:
        vals = ", ".join(["'" + str(e) + "'" for e in empresas])
        f.append(f"EMPRESA IN ({vals})")
    if ejecutivos:
        vals = ", ".join(["'" + str(j) + "'" for j in ejecutivos])
        f.append(f"NOMBRE_EJECUTIVO IN ({vals})")
    return "WHERE " + " AND ".join(f) if f else ""

@router.get("/cumplimiento")
def get_cumplimiento(inicio: Optional[str]=None, fin: Optional[str]=None, codigos: Optional[List[str]]=Query(None), empresas: Optional[List[str]]=Query(None), ejecutivos: Optional[List[str]]=Query(None)):
    conn = sqlite3.connect(DB_PATH)
    where = get_where(inicio, fin, codigos, empresas, ejecutivos)
    sel = ", ".join([f"AVG({v}) as {k}" for k, v in COLS.items()])
    df = pd.read_sql(f"SELECT {sel} FROM ANALISIS_TOK3M {where}", conn)
    conn.close()
    return [{"item": k, "promedio": round(float(df.iloc[0][k] or 0), 1)} for k in COLS.keys()]

@router.get("/evolucion")
def get_evolucion(inicio: Optional[str]=None, fin: Optional[str]=None, codigos: Optional[List[str]]=Query(None), empresas: Optional[List[str]]=Query(None), ejecutivos: Optional[List[str]]=Query(None)):
    conn = sqlite3.connect(DB_PATH)
    where = get_where(inicio, fin, codigos, empresas, ejecutivos)
    sel = ", ".join([f"AVG({v}) as {k}" for k, v in COLS.items()])
    df = pd.read_sql(f"SELECT YMD as fecha, {sel} FROM ANALISIS_TOK3M {where} GROUP BY YMD ORDER BY YMD", conn)
    conn.close()
    return df.to_dict(orient='records')

File: server/routes/calidad.js
const express = require('express');
const router = express.Router();

// FILTRO ULTRA-SEGURO
const getFilters = (req) => {
    const { inicio, fin, empresas, ejecutivos, contactos } = req.query;
    let f = [];
    
    if (inicio && fin && inicio !== '' && fin !== '') {
        f.push(`ymd BETWEEN ${inicio.replace(/-/g, '')} AND ${fin.replace(/-/g, '')}`);
    }

    if (empresas && empresas !== '') {
        f.push(`"EMPRESA" = ANY(string_to_array('${empresas}', ','))`);
    }
    if (ejecutivos && ejecutivos !== '') {
        f.push(`"NOMBRE_EJECUTIVO" = ANY(string_to_array('${ejecutivos}', ','))`);
    }
    if (contactos && contactos !== '') {
        f.push(`"CODIGO_CONTACTO" = ANY(string_to_array('${contactos}', ','))`);
    }

    return f.length > 0 ? ' WHERE ' + f.join(' AND ') : '';
};

const CAL_COLS = ["SALUDO", "TITULAR", "FAMILIAR", "PRESENTACION", "CORDIALIDAD", "RECADO", "EMPEX", "ENCARGO", "GRABADO", "INFORMACION", "MOTIVO", "OFERTA", "CANALES", "COPA", "DUDAS", "CIERRE"];

router.get('/cumplimiento', async (req, res) => {
    try {
        const pool = req.app.get('pool');
        const w = getFilters(req);
        
        const sel = CAL_COLS.map(c => `AVG(COALESCE("${c}", 0)) as "${c}"`).join(',');
        const q = `SELECT ${sel}, AVG(COALESCE("FINAL", 0)) as "FINAL" FROM resumen_maestro ${w}`;
        
        const r = await pool.query(q);
        
        // PARACAÍDAS: Si no hay datos, devolvemos ceros, NUNCA error
        if (!r.rows || r.rows.length === 0 || r.rows[0].FINAL === null) {
            const vacio = CAL_COLS.map(k => ({ item: k, promedio: 0 }));
            vacio.push({ item: "FINAL", promedio: 0 });
            return res.json(vacio);
        }

        const data = r.rows[0];
        const resu = CAL_COLS.map(k => ({ 
            item: k, 
            promedio: Math.round(parseFloat(data[k] || 0) * 10) / 10 
        }));
        resu.push({ item: "FINAL", promedio: Math.round(parseFloat(data.FINAL || 0) * 10) / 10 });
        
        res.json(resu);
    } catch (e) { 
        console.error("ERROR CRÍTICO CALIDAD:", e.message);
        // Devolvemos estructura mínima para que el front NO se ponga blanco
        res.json([]); 
    }
});

router.get('/evolucion', async (req, res) => {
    try {
        const pool = req.app.get('pool');
        const w = getFilters(req);
        
        const sel = CAL_COLS.map(c => `AVG(COALESCE("${c}", 0)) as "${c}"`).join(',');
        const q = `SELECT ymd as fecha, ${sel}, AVG(COALESCE("FINAL", 0)) as "FINAL" 
                  FROM resumen_maestro ${w} 
                  GROUP BY ymd ORDER BY ymd`;
        
        const r = await pool.query(q);
        res.json(r.rows || []);
    } catch (e) { 
        res.json([]); 
    }
});

module.exports = router;